[FLINK-26394] CheckpointCoordinator.isTriggering can not be reset in a particular case.

pltbkd opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

CheckpointCoordinator.isTriggering can not be reset if a checkpoint expires while the checkpointCoordinator task is queuing in the SourceCoordinator executor. Once the issue happens the job can never recover until restarted. This PR is fixing this issue.


## Brief change log

  - Throw timeout exception if the checkpoint completable future expires. The timeout is same as the checkpoint timeout.


## Verifying this change

Tests will be added later.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

pltbkd commented:
This is a preview for discussion. Please do not merge at present.

pltbkd commented:
> @pltbkd Could you please add a ITcase to reproduce this problem?
> 
> BTW, you can tag this PR as `draft` if you think this PR is still under development.

Thanks for the suggestion and sorry for the late reply.

I have added a UT case for CheckpointCoordinator. It would fail without the patch.

becketqin commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -124,6 +124,9 @@
     private final Map<Long, PendingCheckpoint> pendingCheckpoints;
 
+    @GuardedBy("lock")
+    private final Map<Long, CompletableFuture<?>> pendingCheckpointPromises;
```
Could the pending future just be a part of `PendingCheckpoint`?

becketqin commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -654,6 +661,19 @@ private void startTriggeringCheckpoint(CheckpointTriggerRequest request) {
+                                        pendingCheckpoint.getCheckpointID(), checkpointComplete);
+                            } else {
+                                // what TODO?
+                                checkpointComplete.completeExceptionally(new RuntimeException());
```
It looks that this only happens when the pending checkpoint got aborted right after it is created. Should probably use `PendingCheckpoint.getFailureCause()` instead.

becketqin reviewed (commented):
@pltbkd  Thanks for the patch. LGTM overall, just two minor comments.

pltbkd commented:
@becketqin 
Thanks for the review! I've updated the PR fixing these issues.
The future is now added in the constructor of the PendingCheckpoint, and we no longer need to handle the case of the second issue.

becketqin reviewed (approved):
Thanks for the updating the patch. LGTM. Will merge the patch after the CI passes.

pltbkd commented:
I found the CI has failed and I'm finding out the reason.

pltbkd commented:
@flinkbot run azure

pltbkd commented:
@flinkbot run azure

pltbkd commented:
@flinkbot run azure

pltbkd commented:
@flinkbot run azure

becketqin commented:
The CI failure seems irrelevant.  I'll just merge the PR.

becketqin commented:
Merged.
master: 31222b9adf1c354b22fd50c587efc16734b18d40
release-1.15: 9fc89a05f128ab645b73687f240fb14b57790fc6

gezhiwei8899 commented:
![image](https://user-images.githubusercontent.com/16362526/184998667-cad027df-00cf-4e6c-9d28-c9e52ef25626.png)

gezhiwei8899 commented:
Never trigger ckp again

gezhiwei8899 commented:
![image](https://user-images.githubusercontent.com/16362526/185000431-cda381d5-d1ab-4bc5-9a19-686d6c88eb06.png)

pltbkd commented:
@gezhiwei8899 
The commit is revert on 1.15 due to FLINK-27148. Both of the issues are now fixed on the master branch, but are not included in any 1.15 branch. I'm now trying to back port the changes to release-1.13, release-1.14 and release-1.15.

https://github.com/apache/flink/pull/20681
https://github.com/apache/flink/pull/20682
https://github.com/apache/flink/pull/20683

By the way, "Checkpoint Coordinator is suspending" is expected while the job is doing failover, and the coordinator should be restarted along with the job after the failover is done. I suppose it has nothing to do with the issue in this PR.
